proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+That is, the system may choose to migrate thread contexts that arrive at a
+reorder point to different processors, different waves or groups, or indices
+within waves or groups. Such migrations can be visible to the application if
+the application queries certain system state, such as group or thread IDs,
+wave lane indices, ballots, etc.
+
+The existing DXR API has reorder points due to `TraceRay` and `CallShader`.
+The _Shader Execution Reordering_ specification adds several new reorder
+points. The following is a comprehensive list of existing and newly added
+reorder points:
+
+- `TraceRay`: transitions to and from `anyhit`, `intersection`, `closesthit`,
+and `miss` shaders. In the case of multiple `anyhit` or `intersection` shader
+invocations, each shader stage transition is a separate reorder point.
+- `CallShader`: transitions to and from the `callable` shader.
+- `HitObject::TraceRay`: transitions to and from `anyhit` and `intersection`
+shaders. In the case of multiple `anyhit` or `intersection` shader
+invocations, each shader stage transition is a separate reorder point.
+- `HitObject::Invoke`: transitions to and from `closeshit` and `miss` shaders.
+Constitutes a reorder point even in cases where no shader is invoked.
+- `ReorderThread`: the `ReorderThread` call site.
+
+`ReorderThread` stands out as it explicitly separates reordering from a
+transition between shader stages, thus, it allows applications to (carefully)
+choose the most effective reorder locations given a specific workload. The
+combination of `HitObject` and coherence hints provides additional control
+over the reordering itself. These characteristics make `ReorderThread` a
+versatile tool for improving performance in a variety of workloads that suffer
+from divergent execution or data access.
+
+Similar to `TraceRay` and `CallShader`, the behavior of the actual reorder
+operation at any reorder point is implementation specific. An implementation
+may, for instance, not reorder at all, reorder within some local domain
+(threads on a local processor), reorder only within a certain time window,
+reorder globally across the entire GPU and the entire dispatch, or any
+variation thereof. The general target is for implementations to make a _best
+effort_ to extract as much coherence as possible, while keeping the overhead
+of reordering low enough to be practical for use in typical raytracing
+scenarios.
+
+---
+
+### Divergence behavior
+
+An implementation may only reorder threads whose control flow arrives at a
+reorder point. Threads that do not arrive at a reorder point are guaranteed to
+not migrate, even if neighboring threads in the same wave do migrate.
+
+From the physical wave's point of view, an implementation may remove, replace,
+or do nothing with threads that arrive at a reorder point. It is also legal
+for an implementation to replace terminated threads in a wave if any of its
+threads arrive at a reorder point. This may indirectly affect threads that
```

Comment:
It’s possible to add threads even if no threads terminated before. Maybe we can reformulate that to something like
> It is also legal for an implementation to add threads to a wave if any of its threads arrive at a reorder point.

?

One case where threads could be added added without threads terminating first is for odd dispatch sizes. For example if a wave had 32 threads and a dispatch with 48 threads is started:
```hlsl
int MyFunc(int coherenceCoord)
{
    int A = WaveActiveBallot(true);
    // Wave 0: A = 32
    // Wave 1: A = 16
    if (WaveIsFirstLane())
    {
        ReorderThread(coherenceCoord, 32);
        // Possible reordering:
        // Thread 0 from Wave 0 is sorted into Wave 1 as Thread 16
    }
    int B = WaveActiveBallot(true);
    // Wave 0: B = 31
    // Wave 1: B = 17
    // For Wave 1 it appears like a thread was added
    return A - B;
}
```

Reply:
Then again, it doesn't apply to threads that are inactive from control flow. I think a thread that never started the dispatch could be considered to be terminated.

Reply:
I see your point. I think a clearer formulation would be nice. Then it gives less room for misunderstandings.

Without your explanation, I understood it in a way that no new threads can become active (as they can only be removed or replaced), so for a thread that *does not reorder*, I could always `assert(WaveActiveCountBits(true) <= activeThreadsAtShaderStart)`, but that would have been incorrect.

Reply:
This has now been clarified.

Reply:
Thanks!
(I can’t mark this as resolved myself)